packages/peregrine/lib/talons/FilterModal/useFilterModal.js

```diff
@@ -26,32 +31,76 @@ export const useFilterModal = props => {
     const history = useHistory();
     const { pathname, search } = useLocation();
 
-    // iterate over filters once to set up all the collections we need
+    const { data: introspectionData, error: introspectionError } = useQuery(
+        FILTER_INTROSPECTION
+    );
+
+    useEffect(() => {
+        if (introspectionError) {
+            console.error(introspectionError);
+        }
+    }, [introspectionError]);
+
+    // Get "allowed" filters by intersection of aggregation attribute codes and
+    // schema input field types. This restricts the displayed filters to those
+    // that the api will understand.
+    const possibleFilters = useMemo(() => {
+        let nextFilters;
+        if (introspectionData) {
+            nextFilters = introspectionData.__type.inputFields
+                .map(field => field.name)
+                .filter(filterName =>
+                    filters
+                        .map(filter => filter.attribute_code)
+                        .includes(filterName)
```

Comment:
This is a simple functional style, but recall what the original comment here read:

```es6
// iterate over filters once to set up all the collections we need
```

One of the reasons I encourage us to use `for...of` instead of `Array` prototype methods in most places is that it encourages us to combine operations into a single loop. (Another reason is it encourages us to go ahead and use `Set` and `Map` more.) As written, there are a bunch of nested loops happening, which is pretty inefficient. Let's try to improve it.

```es6
const DISABLED = new Set(["category_id", "price"])

const Foo = () => {
  /* ... */
  
  const inputFields = introspectionData
    ? introspectionData.__type.inputFields
    : []

  // precalculate attribute codes, since that's all we need
  // avoid repeating this calculation for each input field
  const attributeCodes = useMemo(
    () => filters.map(({ attribute_code }) => attribute_code)
    [filters]
  )

  // determine which input fields are relevant
  const possibleFilters = useMemo(() => {
    const nextFilters = new Set()

    // perform mapping and filtering in the same cycle
    for (const { name } of inputFields) {
      const isValid = attributeCodes.includes(name)
      const isEnabled = !DISABLED.has(name)

      if (isValid && isEnabled) {
        nextFilters.add(name)
      }
    }

    return nextFilters
  }, [attributeCodes, inputFields])

  /* ... */
}
```

Generally, you can get it down to a couple of loops. Precalculate maps and other structures you can use in the main loop for lookups, then do the bulk of the work in that one main loop.

Also, please restore the original comment above L76. 🙂

Reply:
DONE!